### PR TITLE
8299817: [s390] AES-CTR mode intrinsic fails with multiple short update() calls

### PR DESCRIPTION
**Pull request by RealLucy**
This PR addresses an issue in the AES-CTR mode intrinsic on s390. When a message is ciphered in multiple, small (< 16 bytes) segments, the result is incorrect.

This is not just a band-aid fix. The issue was taken as a chance to restructure the code. though still complicated, It is now easier to read and (hopefully) understand.

Except for the new jetreg test, the changes are purely s390. There are no side effects on other platforms. Issue-specific tests pass. Other tests are in progress. I will update this PR once they are complete. 

**Reviews and comments are very much appreciated.**

@backwaterred could you please run some "official" s390 tests? Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299817](https://bugs.openjdk.org/browse/JDK-8299817): [s390] AES-CTR mode intrinsic fails with multiple short update() calls

### Reviewers
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [ca40de84](https://git.openjdk.org/jdk/pull/11967/files/ca40de8486954ce064f1eb2388fa313cdc3ca1de)
 * @offamitkumar (no known openjdk.org user name / role) ⚠️ Review applies to [0c8ced1e](https://git.openjdk.org/jdk/pull/11967/files/0c8ced1e771d7112cdf263eb23937acbe97ece11)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [0c8ced1e](https://git.openjdk.org/jdk/pull/11967/files/0c8ced1e771d7112cdf263eb23937acbe97ece11)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11967/head:pull/11967` \
`$ git checkout pull/11967`

Update a local copy of the PR: \
`$ git checkout pull/11967` \
`$ git pull https://git.openjdk.org/jdk pull/11967/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11967`

View PR using the GUI difftool: \
`$ git pr show -t 11967`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11967.diff">https://git.openjdk.org/jdk/pull/11967.diff</a>

</details>

### CODE REVIEW

**Comment by offamitkumar** [comment]
Hi @RealLucy , Sorry for kept you waiting. I've run tests over s390/Z machine and everything seems fine (Test failures I'm seeing were already there even before this PR). 

But please let me know if anything specific you want me to test, as of now I've run tier1 test in fast & slow debug build for this PR.

**Comment by RealLucy** [comment]
Thank you for testing, Amit. 
In addition to that, an identical version of the code runs in our internal tests for our commercial product. Everything is fine there as well.

File: test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022 SAP SE. All rights reserved.
+ * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
```
Comment (TheRealMDoerr) [inline test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java:4]: Wrong Copyright header! Please also update the Copyright years.

Comment (RealLucy) [reply test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java]: Sorry for the delay. Copyright is fixed.

**Comment by MBaesken** [comment]
Hi Lutz, is the  JIT_TIMER  still needed ?

**Comment by MBaesken** [comment]
Some small typos found 

nesessary  ->  necessary
sace extended SP ->  save extended SP

Why do you clear memory only in the ASSERT case here ?

src/hotspot/cpu/s390/stubGenerator_s390.cpp

#ifdef ASSERT
    __ clear_mem(Address(Z_SP, (intptr_t)8), resize_len - 8);
#endif

**Comment by RealLucy** [comment]
Typos repaired, JIT_TIMER references removed.
The allocated stack space is cleared for debugging purposes only.
Thanks for having a look, @MBaesken

**Review by MBaesken — APPROVED** [review]

**Comment by RealLucy** [comment]
Being an s390-only PR, failures on arm build and linux-x85 tests are unrelated.

**Review by offamitkumar — APPROVED** [review]
Build and Testing done for tier1 in fastdebug, slowdebug and release build on z15/s390x machine.

 LGTM

File: test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    private static final int LEN_STEPS = 13;
+    private static final int LEN_MAX   = LEN_INC*LEN_STEPS;
+    private static final int SEG_INC   = 3;
+    private static final int SEG_MAX   = 11;
+    private static final int SHOW_ARRAY_LIMIT = 72;
+    private static final boolean DEBUG_MODE = false;
+
+    public static void main(String[] args) throws Exception {
+        if (!DEBUG_MODE) {
+            if (!Compiler.isIntrinsicAvailable(CompilerWhiteBoxTest.COMP_LEVEL_FULL_OPTIMIZATION, "com.sun.crypto.provider.CounterMode", "implCrypt", byte[].class, int.class, int.class, byte[].class, int.class)) {
```
Comment (TheRealMDoerr) [inline test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java:67]: Maybe break the large lines?

**Review by TheRealMDoerr — APPROVED** [review]
Not a detailed review, but I couldn't spot anything bad.

File: test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java (hunk 2 above)
Comment (RealLucy) [reply test/hotspot/jtreg/compiler/codegen/aes/Test8299817.java]: Lines split.
Thank you very much for your reviews, @TheRealMDoerr @offamitkumar @MBaesken

**Comment by RealLucy** [comment]
Build and test failures seem unrelated to this s390-only fix.

**Comment by RealLucy** [comment]
/integrate

All SAP-internal testing is fine.

**Pull request closed without merging**